### PR TITLE
A periodic table instant answer for atomic numbers and masses

### PR DESCRIPTION
**Pull request by zblair**
**What does your Instant Answer do?**
It answers questions like "atomic weight of Mercury" or "atomic mass of Hg" that one might normally use a periodic table for.

**What problem does your Instant Answer solve (Why is it better than organic links)?**
It provides a quick answer to this basic question without the need to click on a Wikipedia link.

**What is the data source for your Instant Answer? (Provide a link if possible)**
http://webelements.com/ and Wikipedia

**Why did you choose this data source?**
They are both popular sources of information online for atomic mass and atomic number information.

**Are there any other alternative (better) data sources?**
There are sources that provide atomic masses as ranges or with more precision, but I thought that this instant answer would provide simpler atomic masses such as those often found on periodic tables.

**What are some example queries that trigger this Instant Answer?**
atomic mass of Nitrogen
atomic mass of N
N atomic mass
atomic number of Nitrogen
atomic number of N
N atomic number

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
High school and university chemistry students

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
Not that I am aware of.

**Which existing Instant Answers will this one supercede/overlap with?**
Some queries (e.g. "atomic mass of N") currently show an answer regarding the definition of "Atomic mass" from Wikipedia. However, longer queries such as "atomic mass of Nitrogen" don't seem to overlap with anything.

**Are you having any problems? Do you need our help with anything?**
I don't think so.

**Where did you hear about DuckDuckHack? (For first time contributors)**
I noticed it in the Menu > Get Involved > Develop link on https://duckduckgo.com/

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
atomic_mass_ddg_instant_answer

## Checklist
Please place an 'X' where appropriate.

```
[X] Added metadata and attribution information (https://duck.co/duckduckhack/metadata)
[X] Wrote test file and added to t/ directory (https://duck.co/duckduckhack/goodie_tests)
[X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[X] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[X] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 10

    - Windows 7
        [X] Google Chrome
        [X] Firefox
        [] Opera
        [] IE 8
        [X] IE 9
        [] IE 10

    - Windows XP
        [] IE 7
        [] IE 8

    - Mac OSX
        [] Google Chrome
        [] Firefox
        [] Opera
        [] Safari

    - iOS (iPhone)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - iOS (iPad)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - Android
        [] Firefox
        [] Native Browser
        [] Google Chrome
        [] Opera
```

### CODE REVIEW

**Comment by zekiel** [comment]
@zblair  thanks for this! Someone will take a look and get back to you asap.

I also linked to this pull request in this idea: https://duck.co/ideas/idea/4728/quick-element-info-periodic-table which seems similar enough.

File: lib/DDG/Goodie/PeriodicTable.pm [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+secondary_example_queries 'atomic weight of Na';
+category 'physical_properties';
+topics 'science';
+code_url 'https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/PeriodicTable.pm';
+attribution github => [ 'zblair', 'Zachary D Blair' ];
+
+my @elements = @{ Load( scalar share('elements.yml')->slurp ) };
+
+# Triggers
+triggers any => 'atomic mass', 'atomic weight', 'atomic number';
```
Comment (mintsoft) [inline lib/DDG/Goodie/PeriodicTable.pm]: What about adding "proton number" and "nucleon number" too?

File: share/goodie/periodic_table/elements.yml [hunk 2]
```diff
@@ -0,0 +1,118 @@
+---
+- [1, "1.008", "Hydrogen", "H"]
```
Comment (mintsoft) [inline share/goodie/periodic_table/elements.yml]: If you changed this to be a hash based on symbol "H" you could avoid the loops in the lookups?

File: lib/DDG/Goodie/PeriodicTable.pm [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    # Look for a matching element in the table
+    my $atomic_number;
+    my $atomic_mass;
+    my $element_name;
+    my $element_symbol;
+ELEMENT_SEARCH: foreach my $element (@elements) {
+
+        # Some elements have multiple names or symbols
+        if ( ref $element->[2] ) {
```
Comment (mintsoft) [inline lib/DDG/Goodie/PeriodicTable.pm]: I think (just from a brief glance) this could be replaced with a simple `grep`: http://perldoc.perl.org/functions/grep.html

File: lib/DDG/Goodie/PeriodicTable.pm (hunk 1 above)
Comment (zblair) [inline lib/DDG/Goodie/PeriodicTable.pm]: Thanks, that's a good idea. I have just committed a change (e3fdf0b5918f1c69715a273f85a55b99fcb38680) that adds "proton number". I didn't add nucleon number because from what I can tell, that's not quite the same as atomic mass, and is different for different isotopes of the same element.

File: lib/DDG/Goodie/PeriodicTable.pm (hunk 3 above)
Comment (zblair) [inline lib/DDG/Goodie/PeriodicTable.pm]: Thanks for the suggestion. I agree, that loop can be simplified. I've tried to simplify the search using List::Util::first, since it's like grep but returns only the first match. I've included that change in my commit, e3fdf0b5918f1c69715a273f85a55b99fcb38680.

File: share/goodie/periodic_table/elements.yml (hunk 2 above)
Comment (zblair) [inline share/goodie/periodic_table/elements.yml]: Thanks for the suggestion. I was thinking of doing something like that, but I am using both the element symbol ("H") and name ("Hydrogen") as keys, depending on the form of the query. I didn't know which one to choose as the key for the hash, so I just went with the list structure instead of a hash.

I could create a hash that maps element symbols to rows in the list, and another hash that maps element names to rows in the list, but I'm not sure if that would be preferable given the greater complexity. If you think that it would be better to have the lookup done using a hash, I'd be happy to do that.

File: lib/DDG/Goodie/PeriodicTable.pm (hunk 1 above)
Comment (mintsoft) [inline lib/DDG/Goodie/PeriodicTable.pm]: Yes, good point! That's very much the case

File: share/goodie/periodic_table/elements.yml (hunk 2 above)
Comment (mintsoft) [inline share/goodie/periodic_table/elements.yml]: Personally I'd prefer hash lookups; I think it'll actually make the code clearer as it'll essentially be reduced to two key lookups. @MrChrisW what'd you think?

Comment (MrChrisW) [inline share/goodie/periodic_table/elements.yml]: Looks like this has now been addressed with e3fdf0b at least the way the lookup is handled using `List::Util` which is fast enough.

**Comment by mintsoft** [comment]
@moollaza @jagtalon ping

**Comment by jagtalon** [comment]
Awesome, checking it out, @zblair!

File: lib/DDG/Goodie/PeriodicTable.pm [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+    # Look for a matching element in the table
+    my $match = first { lc $_->[2] eq $query || lc $_->[3] eq $query } @elements or return;
+    my ( $atomic_number, $atomic_mass, $element_name, $element_symbol ) = @{$match};
+
+    # Return the result if the element was found
+    if ($is_mass_query) {
+        return "$element_name ($element_symbol), Atomic mass $atomic_mass u", structured_answer => {
+            input     => ["$element_name ($element_symbol)"],
+            operation => 'atomic mass',
```
Comment (mintsoft) [inline lib/DDG/Goodie/PeriodicTable.pm]: Can this be switched to title case? i.e. "Atomic Mass"

File: lib/DDG/Goodie/PeriodicTable.pm [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            result    => "$atomic_mass u"
+        };
+    }
+    else {
+        return "$element_name ($element_symbol), Atomic number $atomic_number", structured_answer => {
+            input     => ["$element_name ($element_symbol)"],
+            operation => 'atomic number',
```
Comment (mintsoft) [inline lib/DDG/Goodie/PeriodicTable.pm]: Same again, "Atomic Number"

**Comment by jagtalon** [comment]
@zblair pretty straightforward! Should be good to go. Will just consult @abeyang for a quick review (will put it up on a server in a bit.)

File: lib/DDG/Goodie/PeriodicTable.pm (hunk 5 above)
Comment (jagtalon) [inline lib/DDG/Goodie/PeriodicTable.pm]: Agree--kinda looks better title cased. Prob should do the same with calculator and conversion. Will make an issue for it. :)

**Comment by jagtalon** [comment]
Available to test in https://ddh1.duckduckgo.com/?q=atomic+weight+of+Na&ia=answer

File: t/PeriodicTable.t [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+                input => ['Copernicium (Cn)'],
+                operation => 'atomic number',
+                result => '112'
+            },
+        ),
+    'atomic number of Uub' => test_zci(
+            'Ununbium (Uub), Atomic number 112',
+            structured_answer => {
+                input => ['Ununbium (Uub)'],
+                operation => 'atomic number',
```
Comment (jagtalon) [inline t/PeriodicTable.t]: You'd have to update the tests too for the case. :)

**Comment by abeyang** [comment]
Hi @zblair, looking good! However, it would be nice to add a little pizazz to this:

atomic-weight-v1-alkali
atomic-weight-v1-metals
atomic-weight-v1-noble_gas
atomic-weight-v1-nonmetals

I've broken the periodic table down to four categories (I know there are more, but we're going with a minimal color palette for now): 

- $green: Metalloids + Nonmetals (except noble gases)
- $blue-light: Noble Gases
- $gold: Alkali metals + Alkaline earth metals
- $red: All other metals (including transition metals)

Basically, this is what the periodic table would look like, with these colors:

periodic-table

Let us know if you have any questions, or if you need help with anything. Looking forward to seeing this go live!

**Comment by MrChrisW** [comment]
@abeyang Damn that's sexy :grinning: Can't wait to see this live :+1:

**Comment by jagtalon** [comment]
@abeyang That looks super cool!

**Comment by jagtalon** [comment]
@zblair this is good for now. We can do https://github.com/duckduckgo/zeroclickinfo-goodies/pull/892#issuecomment-69277697 on the second iteration. :+1:

**Comment by abeyang** [comment]
hi @zblair, after talking it over with some folks, I think we're good to go to release what you have here as a v1, and hold off the newer designs for the future as a v2. Thanks!

**Comment by jagtalon** [comment]
Thanks @abeyang!

**Comment by jagtalon** [comment]
@zblair it's live! https://duckduckgo.com/?q=atomic+weight+of+cobalt&ia=answer

**Comment by zblair** [comment]
@jagtalon, Awesome! Thanks for letting me know.

**Comment by moollaza** [comment]
@zblair Now that you've completed your first Goodie, we really hope you'll work on another! Feel free to check out our [Ideas Forum](https://duck.co/ideas) or look through our Github issues if you need something to work on!
